cabal-install/src/Distribution/Client/ProjectBuilding.hs

```diff
@@ -637,6 +646,32 @@ rebuildTargets verbosity
                           , elabSetupPackageDBStack elab ]
         ]
 
+    offlineError :: BuildOutcomes
+    offlineError = Map.fromList . map makeBuildOutcome $ packagesToDownload
+      where
+        makeBuildOutcome :: ElaboratedConfiguredPackage -> (UnitId, BuildOutcome)
+        makeBuildOutcome ElaboratedConfiguredPackage {
+          elabUnitId,
+          elabPkgSourceId = PackageIdentifier { pkgName, pkgVersion }
+        } = (elabUnitId, Left (BuildFailure {
+          buildFailureLogFile = Nothing,
+          buildFailureReason = DownloadFailed . error $ makeError pkgName pkgVersion
```

Comment:
I think it shouldn't involve `error`: it's not an exceptional situation like a network failure, or something else out of our control, it's we who are refusing to do something because of the flag -- it's no exception! So, I think, it should be `DependentFailed` rather than `DownloadFailed`. Perhaps someone has another opinion.

Reply:
`DependentFailed` needs another package as a parameter this looks like it's used to say that the package that is failing has a dependency that failed. I need to say that this specific package failed while also not producing an exception. Nothing else in the `BuildFailureReason` data type looked like it suited the problem so I added a `GracefulFailure` option to `BuildFailureReason` and it takes a string that will be printed to the terminal without any extra text or callstack, I think anything more specific would clutter up the data type more than is needed.

Reply:
Full disclosure: I'm no expert in Cabal error types. But, I thought that `DependentFailed` would work quite all right for this use case. Imagine you depend on 10 packages and 9 of them are in the cache. Then, you may want to know which one wasn't there. On the other hand, if there're two dependencies that are not in the cache then reporting just one is not so nice, perhaps. Although, I could live with reporting one that happened to be the first to be required during the build…

That said, I don't have a strong preference. I'm just not a big fan of changing data types that sound fundamental. Cabal is a library and this would be a breaking change...